FINCN-252 Machine Learning Scorecard for Credit Risk Assessment Phase 4

xurror opened the pull request:
## Description

Credit Scorecard Algorithms integrations with fineract backend

https://issues.apache.org/jira/browse/FINCN-252

awasum commented:
Thanks very much for this work. Why is the issue on the Fineract CN issue tracker and not on the Fineract 1.x issue tracker? cc @edcable

xurror commented:
I think there was the idea of making it a microservice, that's why we have a seperate repository where the module server is hosted so it can be run independently.
Some changes feature here because the first level of integration of the project was focused on it's usability with Fineract 1.x.

Referencing Here the UI level changes(https://github.com/openMF/web-app/pull/1353) and Scorecard Server(https://github.com/apache/fineract-credit-scorecard)

xurror commented:
I've finally added the necessary IT tests for this module.

xurror committed:
FINCN-351 Make the scorecard integration into a plugin

  - Credit Scorecard module:

     - Credit Scorecard organisational module API
     - Rule Based and ML Scoring Entities
     - Credit Scorecard Validators and JSON assemblers
     - Credit Scorecard Read and Write Services
     - Credit Scorecard Flyway Migration
     - Credit scorecard Loan Product Feature Configurations
     - Credit Scorecard integration in Loan Account Creation
     - FINCN-331 Statistical Credit Scorecard

ptuomola commented:
Looks interesting! Can you fix the merge conflicts and squash the commits please, and I'll take a look...

awasum commented:
@xurror Please fix merge conflicts

xurror commented:
The PR is good now. Please review when you can. @ptuomola @awasum

BLasan commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/creditscorecard/domain/CreditScorecard.java:
```diff
@@ -0,0 +1,115 @@
+        return scoringModel;
+    }
+
+    public org.apache.fineract.portfolio.creditscorecard.domain.RuleBasedScorecard getRuleBasedScorecard() {
```
Can't we import this and directly use `RuleBasedScorecard`?

BLasan commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/creditscorecard/domain/CreditScorecardFeature.java:
```diff
@@ -0,0 +1,166 @@
+
+    public static CreditScorecardFeature fromJson(JsonCommand command) {
+        final String name = command.stringValueOfParameterNamed("name");
+        final org.apache.fineract.portfolio.creditscorecard.domain.FeatureValueType valueType = org.apache.fineract.portfolio.creditscorecard.domain.FeatureValueType
```
Here too

BLasan commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/creditscorecard/domain/CreditScorecardFeature.java:
```diff
@@ -0,0 +1,166 @@
+        final EnumOptionData dataType = CreditScorecardEnumerations.featureDataType(this.dataType);
+        final EnumOptionData category = CreditScorecardEnumerations.featureCategory(this.category);
+
+        return CreditScorecardFeatureData.instance(null, getId(), this.name, valueType, dataType, category, this.active, null, null, null,
```
Can't we write a overloaded method for `CreditScorecardFeatureData.instance` instead of passing many `null`s? Just an idea :)

BLasan commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/creditscorecard/domain/MLScorecard.java:
```diff
@@ -0,0 +1,136 @@
+    @Column(name = "prediction_request_id")
+    private Integer predictionRequestId;
+
+    public MLScorecard() {
```
Do we need this constructor?

BLasan commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanReadPlatformServiceImpl.java:
```diff
@@ -662,7 +667,12 @@ public String loanSchema() {
+                    + " topup.topup_amount as topupAmount, "
+
+                    + " csc.id as scorecardId, csc.scorecard_scoring_method as scoringMethod, csc.scorecard_scoring_model as scoringModel "
+
```
Unnecessary lines

BLasan commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanReadPlatformServiceImpl.java:
```diff
@@ -681,7 +691,9 @@ public String loanSchema() {
                     + " left join m_loan_topup as topup on l.id = topup.loan_id"
-                    + " left join m_loan as topuploan on topuploan.id = topup.closure_loan_id";
+                    + " left join m_loan as topuploan on topuploan.id = topup.closure_loan_id"
+
```
Here too

BLasan reviewed (changes requested):
LGTM ! Please consider the review comments

awasum commented:
@xurror , please fix review comments if possible, fix conflicts so this can be reviewed.

vidakovic commented:
Closing because it looks abandoned.

awasum commented:
> Closing because it looks abandoned.

I see a commit was made by author 25 days ago.. 
Please @xurror please can you find time to make the build pass and the merge conflicts all fixed if any... It takes  30 days for the bug to be ato closed by the bot.

xurror commented:
I think I've now resolved all conflicts on this PR.

ikimbra commented:
I'm interested in this feature as we're working on the credit scoring tool at https://github.com/lalitsanagavarapu/mifos-scorecard. @francisguchie, and @rrpawar96 what's your take on this?

xurror commented:
Passing the build now.

hafizsheetab commented:
Is there any work going on with this PR or any other repos that have successfully implemented this plugin?
